### PR TITLE
Refactor ObjectMetadata to JSON conversion functions.

### PR DESCRIPTION
**Pull request by coryan**
I need to clean these up to properly support `set_storage_class()`. The
functions had different naming conventions, return types, and fairly
limited testing. Fixed all that so when we implement
`set_storage_class()` the values are inserted in the right calls.

This is part of the work for #2322.

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/2330)

### CODE REVIEW

File: google/cloud/storage/object_metadata_test.cc [hunk 1]
```diff
@@ -183,12 +183,159 @@ TEST(ObjectMetadataTest, IOStream) {
   EXPECT_THAT(actual, HasSubstr("temporary_hold=true"));
 }
 
-/// @test Verify we can convert a ObjectMetadata object to a JSON string.
-TEST(ObjectMetadataTest, UpdatePayload) {
+/// @test Verify that ObjectMetadataJsonForCompose works as expected.
+TEST(ObjectMetadataTest, JsonForComposeEmpty) {
+  internal::nl::json actual = ObjectMetadataJsonForCopy(ObjectMetadata());
```
Comment (devbww) [inline google/cloud/storage/object_metadata_test.cc]: s/Copy/Compose/?

File: google/cloud/storage/object_metadata_test.cc [hunk 2]
```diff
@@ -183,12 +183,159 @@ TEST(ObjectMetadataTest, IOStream) {
...
+  internal::nl::json expected({});
+  EXPECT_EQ(expected, actual);
+}
+
+/// @test Verify the ObjectMetadataJsonForCompose() works as expected.
```
Comment (devbww) [inline google/cloud/storage/object_metadata_test.cc]: s/the/that/

File: google/cloud/storage/object_metadata_test.cc [hunk 3]
```diff
@@ -183,12 +183,159 @@ TEST(ObjectMetadataTest, IOStream) {
...
+}
+
+/// @test Verify that ObjectMetadataJsonForCopy works as expected.
+TEST(ObjectMetadataTest, JsonForCopyEmpty) {
+  internal::nl::json actual = ObjectMetadataJsonForCopy(ObjectMetadata());
+  internal::nl::json expected({});
+  EXPECT_EQ(expected, actual);
+}
+
+/// @test Verify the ObjectMetadataJsonForCopy() works as expected.
```
Comment (devbww) [inline google/cloud/storage/object_metadata_test.cc]: s/the/that/

File: google/cloud/storage/object_metadata_test.cc [hunk 4]
```diff
@@ -183,12 +183,159 @@ TEST(ObjectMetadataTest, IOStream) {
...
+}
+
+/// @test Verify that ObjectMetadataJsonForInsert works as expected.
+TEST(ObjectMetadataTest, JsonForInsertEmpty) {
+  internal::nl::json actual = ObjectMetadataJsonForInsert(ObjectMetadata());
+  internal::nl::json expected({});
+  EXPECT_EQ(expected, actual);
+}
+
+/// @test Verify the ObjectMetadataJsonForInsert() works as expected.
```
Comment (devbww) [inline google/cloud/storage/object_metadata_test.cc]: s/the/that/

File: google/cloud/storage/object_metadata_test.cc [hunk 5]
```diff
@@ -183,12 +183,159 @@ TEST(ObjectMetadataTest, IOStream) {
...
+}
+
+/// @test Verify that ObjectMetadataJsonForRewrite works as expected.
+TEST(ObjectMetadataTest, JsonForRewriteEmpty) {
+  internal::nl::json actual = ObjectMetadataJsonForRewrite(ObjectMetadata());
+  internal::nl::json expected({});
+  EXPECT_EQ(expected, actual);
+}
+
+/// @test Verify the ObjectMetadataJsonForRewrite() works as expected.
```
Comment (devbww) [inline google/cloud/storage/object_metadata_test.cc]: s/the/that/

File: google/cloud/storage/internal/object_requests.cc [hunk 6]
```diff
@@ -122,7 +122,12 @@ internal::nl::json ObjectMetadataJsonForUpdate(ObjectMetadata const& meta) {
   SetIfNotEmpty(metadata_as_json, "contentLanguage", meta.content_language());
   SetIfNotEmpty(metadata_as_json, "contentType", meta.content_type());
 
-  metadata_as_json["eventBasedHold"] = meta.event_based_hold();
+  if (meta.event_based_hold()) {
+    metadata_as_json["eventBasedHold"] = true;
```
Comment (devbww) [inline google/cloud/storage/internal/object_requests.cc]: Might it be worth growing a `SetIfTrue()` to match `SetIfNotEmpty()`?

File: google/cloud/storage/object_metadata_test.cc (hunk 1 above)
Comment (coryan) [reply google/cloud/storage/object_metadata_test.cc]: Yup, thanks!

File: google/cloud/storage/object_metadata_test.cc (hunk 2 above)
Comment (coryan) [reply google/cloud/storage/object_metadata_test.cc]: Fixed.

File: google/cloud/storage/object_metadata_test.cc (hunk 3 above)
Comment (coryan) [reply google/cloud/storage/object_metadata_test.cc]: Fixed.

File: google/cloud/storage/object_metadata_test.cc (hunk 4 above)
Comment (coryan) [reply google/cloud/storage/object_metadata_test.cc]: Fixed.

File: google/cloud/storage/object_metadata_test.cc (hunk 5 above)
Comment (coryan) [reply google/cloud/storage/object_metadata_test.cc]: Fixed.

File: google/cloud/storage/internal/object_requests.cc (hunk 6 above)
Comment (coryan) [reply google/cloud/storage/internal/object_requests.cc]: I looked around and could not find any other usage.  I am going to leave as-is for now.

**Comment by coryan** [comment]
PTAL.

**Pull request merged**